taichi/backends/cuda/cuda_profiler.h

```diff
@@ -1,40 +1,72 @@
 #pragma once
 
+#include "taichi/system/timeline.h"
 #include "taichi/program/kernel_profiler.h"
 #include "taichi/backends/cuda/cuda_driver.h"
 #include "taichi/backends/cuda/cuda_context.h"
+#include "taichi/backends/cuda/cupti_toolkit.h"
 
 #include <string>
 #include <stdint.h>
 
 TLANG_NAMESPACE_BEGIN
 
-struct CUDAEventRecord {
-  std::string name;
-  float kernel_elapsed_time_in_ms{0.0};
-  float time_since_base{0.0};
-  void *start_event{nullptr};
-  void *stop_event{nullptr};
+enum class ProfilingToolkit : int {
+  event,
+  cupti,
 };
 
-// A CUDA kernel profiler that uses CUDA timing events
+class EventToolkit;
+
+// A CUDA kernel profiler
 class KernelProfilerCUDA : public KernelProfilerBase {
  public:
-  std::string title() const override;
-  KernelProfilerBase::TaskHandle start_with_handle(
-      const std::string &kernel_name) override;
-  void record(KernelProfilerBase::TaskHandle &task_handle,
-              const std::string &task_name) override;
+  KernelProfilerCUDA();
+  std::string title() const override {
+    return "CUDA Profiler";
+  }
+
+  void trace(KernelProfilerBase::TaskHandle &task_handle,
+             const std::string &task_name) override;
   void sync() override;
   void print() override;
   void clear() override;
   void stop(KernelProfilerBase::TaskHandle handle) override;
 
+  bool statistics_on_traced_records();
+
+  KernelProfilerBase::TaskHandle start_with_handle(
+      const std::string &kernel_name) override;
+
  private:
-  void *base_event_{nullptr};
+  ProfilingToolkit tool_ = ProfilingToolkit::event;
```

Comment:
nit: will be nice to add a note to highlight an invariant here:
tool_ == event -> event_toolkit_ not nullptr
tool_ == event-> cupti_toolkit_ not nullptr.